### PR TITLE
Feature/forms 679 archive files in go

### PR DESCRIPTION
**Pull request by jekuaitk**
https://jira.itkdev.dk/browse/FORMS-679

* Added functionality allowing files on submissions to be archived in GO.

### CODE REVIEW

File: web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php [hunk 1]
```diff
@@ -296,33 +310,117 @@ private function createSubCase(string $caseId, string $caseName) {
...
+   * Returns array of file elements keys in submission.
+   */
+  private function getFileElementKeysFromSubmission(WebformSubmission $submission): array {
+    $elements = $submission->getWebform()->getElementsDecodedAndFlattened();
+
+    $fileElements = [];
+
+    foreach (self::FILE_ELEMENT_TYPES as $fileElementType) {
+      $fileElements = array_merge($fileElements, $this->getAvailableElementsByType($fileElementType, $elements));
+    }
```
Comment (rimi-itk) [inline web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php:406]: https://dev.to/klnjmm/never-use-arraymerge-in-a-for-loop-in-php-5go1

File: web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php [hunk 2]
```diff
@@ -296,33 +310,117 @@ private function createSubCase(string $caseId, string $caseName) {
...
+
+    foreach ($elementKeys as $elementKey) {
+      if (empty($submission->getData()[$elementKey])) {
+        continue;
+      }
+
+      // Convert occurrences of singular file into array.
+      $elementFileIds = (array) $submission->getData()[$elementKey];
+
+      $fileIds = array_merge($fileIds, $elementFileIds);
```
Comment (rimi-itk) [inline web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php:420]: https://dev.to/klnjmm/never-use-arraymerge-in-a-for-loop-in-php-5go1

File: web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php [hunk 3]
```diff
@@ -296,33 +310,117 @@ private function createSubCase(string $caseId, string $caseName) {
...
+
+    // Handle attached files.
+    if ($shouldArchiveFiles) {
+      $fileIds = $this->getFileElementKeysFromSubmission($submission);
+
+      $childDocumentIds = [];
+
+      foreach ($fileIds as $fileId) {
+        /** @var \Drupal\file\Entity\File $file */
+        $file = $this->entityTypeManager->getStorage('file')->load($fileId);
```
Comment (rimi-itk) [inline web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php:339]: Consider getting the file storage outside the loop.

File: web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php [hunk 4]
```diff
@@ -296,33 +310,117 @@ private function createSubCase(string $caseId, string $caseName) {
...
+
+        $fileContent = file_get_contents($file->getFileUri());
+        $getOrganizedFileName = $webformLabel . '-' . $submission->serial() . '-' . $file->getFilename();
+
+        $childDocumentId = $this->archiveDocumentToGetOrganizedCase($caseId, $getOrganizedFileName, $fileContent);
+
+        $childDocumentIds[] = $childDocumentId;
+      }
+
+      $documentIdsForFinalizing = array_merge($documentIdsForFinalizing, $childDocumentIds);
```
Comment (rimi-itk) [inline web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php:349]: https://dev.to/klnjmm/never-use-arraymerge-in-a-for-loop-in-php-5go1

**Review by rimi-itk — CHANGES_REQUESTED** [review]

File: web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php (hunk 1 above)
Comment (jekuaitk) [reply web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php:406]: Fixed.

File: web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php (hunk 4 above)
Comment (jekuaitk) [reply web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php:349]: This is not inside a for loop.

File: web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php (hunk 3 above)
Comment (jekuaitk) [reply web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php:339]: Moved obtaining of the file storage outside the loop.

File: web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php [hunk 5]
```diff
@@ -402,9 +404,12 @@ private function getFileElementKeysFromSubmission(WebformSubmission $submission)
     $fileElements = [];
 
     foreach (self::FILE_ELEMENT_TYPES as $fileElementType) {
-      $fileElements = array_merge($fileElements, $this->getAvailableElementsByType($fileElementType, $elements));
+      $fileElements[] = $this->getAvailableElementsByType($fileElementType, $elements);
     }
 
+    // https://dev.to/klnjmm/never-use-arraymerge-in-a-for-loop-in-php-5go1
+    $fileElements = array_merge([], ...$fileElements);
```
Comment (rimi-itk) [inline web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php:411]: There's no reason to add `[]`, just use `array_merge(...)` like you use to do.

File: web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php [hunk 6]
```diff
@@ -417,10 +422,10 @@ private function getFileElementKeysFromSubmission(WebformSubmission $submission)
       // Convert occurrences of singular file into array.
       $elementFileIds = (array) $submission->getData()[$elementKey];
 
-      $fileIds = array_merge($fileIds, $elementFileIds);
+      $fileIds[] = $elementFileIds;
     }
 
-    return $fileIds;
+    return array_merge([], ...$fileIds);
```
Comment (rimi-itk) [inline web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php:428]: There's no reason to add `[]`, just use `array_merge(...)` like you use to do.

**Review by rimi-itk — CHANGES_REQUESTED** [review]

File: web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php (hunk 6 above)
Comment (jekuaitk) [reply web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php]: Fixed.

File: web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php (hunk 5 above)
Comment (jekuaitk) [reply web/modules/custom/os2forms_get_organized/src/Helper/ArchiveHelper.php]: Fixed.

**Review by rimi-itk — APPROVED** [review]

**Pull request merged**